x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,114 @@
+        this.migrationByVersion = migrationByVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(task, params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(AllocatedPersistentTask task, int currentMigrationVersion, ActionListener<Void> listener) {
+        if (task.isCancelled()) {
+            listener.onFailure(new TaskCancelledException("Security migration task cancelled"));
+            return;
+        }
+        if (migrationByVersion.lastKey() > currentMigrationVersion) {
+            Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = migrationByVersion.higherEntry(
+                currentMigrationVersion
+            );
+            int clusterMappingsVersion = loadIndexMappingVersion(securityIndexManager.aliasName(), clusterService.state());
+            SecurityMigrations.SecurityMigration migration = migrationEntry.getValue();
+
+            // Check if all nodes can support feature and that the cluster is on a compatible mapping version
+            if (isEligibleSecurityMigration(clusterMappingsVersion, featureService, clusterService.state(), migration)) {
+                migration.migrate(
+                    securityIndexManager,
+                    client,
+                    ActionListener.wrap(
+                        response -> updateMigrationVersion(
+                            migrationEntry.getKey(),
+                            securityIndexManager.getConcreteIndexName(),
+                            ActionListener.wrap(
+                                updateResponse -> applyOutstandingMigrations(task, migrationEntry.getKey(), listener),
+                                listener::onFailure
+                            )
+                        ),
+                        listener::onFailure
+                    )
+                );
+                return;
+            }
+        }
+        logger.info("Security migrations applied until version: [" + currentMigrationVersion + "]");
+        listener.onResponse(null);
+    }
+
+    private void updateMigrationVersion(int migrationVersion, String indexName, ActionListener<Void> listener) {
+        client.execute(
+            UpdateIndexMigrationVersionAction.INSTANCE,
+            new UpdateIndexMigrationVersionAction.Request(TimeValue.MAX_VALUE, migrationVersion, indexName),
+            ActionListener.wrap((response) -> {
+                if (response.isAcknowledged()) {
+                    listener.onResponse(null);
+                } else {
+                    listener.onFailure(new IllegalStateException("Failed to apply new migration version"));
```

Comment:
I don't think we need to fail unacknowledged master actions. So actually maybe not use acknowledgeable ones.
Acknowledged `false`, just means that some nodes failed to apply the update, during some deadline (by reading the comments), but the new state has been committed.

Reply:
Ah nice catch. Will remove it.